### ISSUE TITLE
Dependency analysis: Unnecessary/undesired compilation unit included in executable

### ISSUE THREAD

**Issue by mjambon** [ocaml/dune#753]
There seems to be the following problem:
```
open Hellolib;;
Hello.hello ()  (* = Hellolib.Hello.hello () *)
```
results in any local `hello.ml` to be compiled and linked against our executable, even though we don't use this `Hello` in this particular program.

Suggested workaround: Rename the local `hello.ml`, e.g. `hello_main.ml`.

See https://github.com/mjambon/dune-dependencies for exact reproduction details.

**Comment by diml** [comment]
Yh, unfortunately that's a limitation of using ocamldep. I suppose we could try codept

**Comment by gares** [comment]
May I ask if it is really a limitation of ocamldep or it is the way/time dune invokes it that makes the dependencies analysis incorrect? Because I've never seen this problem before using make and ocamldep by hand. Was I just lucky?

**Comment by gares** [comment]
Letting me write down the dependencies by hand would work for me

**Comment by diml** [comment]
That's exactly the idea of explicit imports.

**Comment by diml** [comment]
BTW, here is a temporary workaround:

```ocaml
module Hello = struct end
open Hellolib
Hello.hello ()
```

**Comment by gares** [comment]
I've checked the doc of coqdep and indeed it says that -module does not resolve dependencies to files. To me this means that the output of -module is nonsensical, it does not respect the semantics of the language wrt binders.

I understand that dune does quite a few renaming behind the scenes, so "resolving to files" may be quite hard in that case but should be still possible (after the various preprocessing business, you could always to back to the orignal names for the sake of computing dependencies). Or one could just improve coqdep and let it accept a "binding map" (eg module X is in file x.pp.ml, rather than x.ml). Am I missing something? (I'm quite a newbie to dune).

As I said before, for lpcic/elpi#24 that is just a few files I'm OK with writing the dependencies by hand.
But I'm not so sure this solution can scale if you have many files.

**Comment by diml** [comment]
I don't believe this is related to the use `-modules`, I get the same behaviour without it:

```ocaml
$ cat foobar.ml 
open Hellolib;;
Hello.hello ()
$ touch hello.ml
$ ocamldep foobar.ml 
foobar.cmo : hello.cmo
foobar.cmx : hello.cmx
```

**Comment by Octachron** [comment]
This is not related to `-modules`: the  main differences between `-modules` and the standard mode of ocamldep in this respect is that the standard mode has a postprocessing filter that only keep dependencies on files that appear in the included directories.

**Comment by rgrinberg** [comment]
> As I said before, for LPCIC/elpi#24 that is just a few files I'm OK with writing the dependencies by hand.
But I'm not so sure this solution can scale if you have many files.

Indeed it's quite an annoying issue, but it's fairly easy to avoid in practice. Use the hack Jeremie suggested, or simply don't define local modules that are named after libraries. Note that if you're writing a library, this should not force you to change your interface. As you can always create a proper name in the lib interface file.

E.g.

```
(library
 (name bar)
 (libraries foo)) ;; now we can't define foo.ml
```

But if we have:
```
$ ls
  bar.ml
  foo0.ml
$ cat bar.ml
module Foo = Foo0
```

All is well again.

**Comment by diml** [comment]
@gares, what problem are you seeing? Do you get extra modules linked in your executables, or does the compilation just fails because of circular dependencies?

**Comment by gares** [comment]
I was getting circular dependencies. Then with some fiddling (that I can't recall exactly, I was trying to understand what was provoking the cycle) I was getting a wrong order of modules being linked in the cmxa that in turn makes all the code linking the cmxa complain that some modules are not defined.
